sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryRelation.scala

```diff
@@ -225,10 +231,31 @@ case class CachedRDDBuilder(
     _cachedColumnBuffers
   }
 
+  def manualClose[T <: CachedBatch](entry: T): T = {
+    val entryManualCloseTasks = Future {
+      entry match {
+        case o: AutoCloseable =>
+          try {
+            o.close
```

Comment:
We can't reuse the manualClose method defined in your previous PR?
I'm also not seeing that CachedBatch implementations have any resources to free - what are you looking at there?
In this case, maybe we don't want to even create the Future unless there is something AutoCloseable to close.

Reply:
@srowen , the reason I also want to do manual close in InMemoryRelation is because since 3.1.1, this class supports user-defined Serializer through "spark.sql.cache.serializer" , and by this new config, user is able to create user-defined CachedBatch and store it inside InMemoryRelation. In our case, we want to define an Apache ArrowCachedBatch to store ColumnarBatch data inside memory or serialize to disk. 
That is why I am also thinking to add a manual close function here , and use AutoCloseable to make it more general.

The reason of adding future here, is because from my local test, releasing couple Gigabytes (ex:  10G) data synchronizely will still hold df.unpersist() for a while.
I also added an UT to explain my purpose with user-define CachedBatchSerializer and CachedBatch.

Reply:
I'm not arguing against Future - just arguing that it's pointless to make a Future if there is nothing to close. The code could be rearranged (here and in the previous commit) to maybe take advantage of that? not sure. Also see comment in your other PR about a thread pool, perhaps.

Such an impl should be implementing finalize of course but I do agree it's OK to try to proactively close.